src/service/video/docking.js

```diff
@@ -0,0 +1,1268 @@
+    // Upcoming fixes: #14657, #14658.
+    // TODO(alanorozco): Cleanup markup for readability once fixes land.
+        htmlFor(this.getDoc_())`
+          <div class="amp-docked-video-controls" hidden>
+            <div class="amp-docked-video-button-group">
+              <div role="button" ref="playButton"
+                  class="amp-docked-video-play"></div>
+              <div role="button" ref="pauseButton"
+                  class="amp-docked-video-pause"></div>
+            </div>
+            <div class="amp-docked-video-button-group">
+              <div role="button" ref="muteButton"
+                class="amp-docked-video-mute"></div>
+              <div role="button" ref="unmuteButton"
+                  class="amp-docked-video-unmute"></div>
+            </div>
+            <div class="amp-docked-video-button-group">
+              <div role="button" ref="fullscreenButton"
+                  class="amp-docked-video-fullscreen"></div>
+            </div>
+            <div class="amp-docked-video-button-group
+                amp-docked-video-dismiss-group">
+              <div role="button" ref="dismissButton"
+                  class="amp-docked-video-dismiss"></div>
+            </div>
+          </div>`)));
+
+    /** @private {?VideoOrBaseElementDef} */
+    this.lastDismissed_ = null;
+
+    /** @private {?RelativeY} */
+    this.lastDismissedPosY_ = null;
+
+    /**
+     * Unlisteners for the currently minimized video.
+     * @private {!Array<!UnlistenDef>}
+     */
+    this.videoUnlisteners_ = [];
+
+    /**
+     * Memoizes x, y and scale to prevent useless mutations.
+     * @private {?{x: number, y: number, scale: number}}
+     */
+    this.placedAt_ = null;
+
+    /** @private {?Direction} */
+    this.scrollDirection_ = null;
+
+    /** @private {number} */
+    this.lastScrollTop_ = this.viewport_.getScrollTop();
+
+    /** @private {boolean} */
+    this.stickyControls_ = false;
+
+    /** @private @const {!function()} */
+    // Lazily invoked.
+    this.install_ = once(() => {
+      // It would be nice if the viewport service provided scroll direction
+      // and speed.
+      this.viewport_.onScroll(
+          throttle(ampdoc.win, () => this.updateScroll_(), 50));
+
+      this.installStyles_();
+    });
+  }
+
+  /** @private */
+  installStyles_() {
+    installStylesForDoc(
+        this.ampdoc_,
+        cssText,
+        /* callback */ null,
+        /* opt_isRuntimeCss */ false,
+        /* opt_ext */ 'amp-video-docking');
+  }
+
+  /** @param {!VideoOrBaseElementDef} video */
+  register(video) {
+    const {element} = video;
+    const fidelity = PositionObserverFidelity.HIGH;
+
+    this.install_();
+
+    this.getPositionObserver_().observe(element, fidelity,
+        throttle(this.ampdoc_.win, () => this.onPositionChange_(video), 40));
+  }
+
+  /** @private */
+  updateScroll_() {
+    const scrollTop = this.viewport_.getScrollTop();
```

Comment:
Unfortunately that doesn't seem to work. It relies on scroll velocity, which is almost always 0, which means that in most cases it will default to "scroll down", even if that's inaccurate.